The data we will use on this module
"wget https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv"

In [1]:
#importing lib's
import pandas as pd 
import numpy as np

In [2]:
url = 'https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv'
df = pd.read_csv(url)
df.head()

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                  170            3.0       159.0     3413.433759   
1                  130            5.0        97.0     3149.664934   
2                  170            NaN        78.0     3079.038997   
3                  220            4.0         NaN     2542.392402   
4                  210            1.0       140.0     3460.870990   

   acceleration  model_year  origin fuel_type         drivetrain  num_doors  \
0          17.7        2003  Europe  Gasoline    All-wheel drive        0.0   
1          17.8        2007     USA  Gasoline  Front-wheel drive        0.0   
2          15.1        2018  Europe  Gasoline  Front-wheel drive        0.0   
3          20.2        2009     USA    Diesel    All-wheel drive        2.0   
4          14.4        2009  Europe  Gasoline    All-wheel drive        2.0   

   fuel_efficiency_mpg  
0            13.231729  
1            13.688217  
2            14.246341  
3            16.912736  
4            12.488369

Q1. Pandas version

In [3]:
pd.__version__

'2.3.1'

Q2. Records count

In [4]:
len(df)

9704

Q3. Fuel types

In [5]:
df['fuel_type'].nunique()

2

Q4. Missing values

In [6]:
df.isnull().sum()

engine_displacement      0
num_cylinders          482
horsepower             708
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64

In [7]:
df.count()

engine_displacement    9704
num_cylinders          9222
horsepower             8996
vehicle_weight         9704
acceleration           8774
model_year             9704
origin                 9704
fuel_type              9704
drivetrain             9704
num_doors              9202
fuel_efficiency_mpg    9704
dtype: int64

we can see there are 4 columns with missing values 

Q5. Max fuel efficiency

In [8]:
df.groupby('origin')['fuel_efficiency_mpg'].max()

origin
Asia      23.759123
Europe    25.967222
USA       24.971452
Name: fuel_efficiency_mpg, dtype: float64

In [9]:
df[df['origin'] == 'Asia']['fuel_efficiency_mpg'].max()

np.float64(23.759122836520497)

What's the maximum fuel efficiency of cars from Asia? we can the max value from asia is 23.759123

Q6. Median value of horsepower

In [10]:
df['horsepower'].median()

np.float64(149.0)

In [11]:
mode_value = df['horsepower'].mode()[0]
mode_value

np.float64(152.0)

In [12]:
df['horsepower'] = df['horsepower'].fillna(mode_value)

In [14]:
df['horsepower']

0       159.0
1        97.0
2        78.0
3       152.0
4       140.0
        ...  
9699    164.0
9700    154.0
9701    138.0
9702    177.0
9703    140.0
Name: horsepower, Length: 9704, dtype: float64

In [15]:
df['horsepower'].median()

np.float64(152.0)

we can see that the value of the median increased 

Q7. Sum of weights


In [16]:

X= df.loc[df['origin'] == 'Asia', ['vehicle_weight', 'model_year']].head().to_numpy()
X

array([[2714.21930965, 2016.        ],
       [2783.86897424, 2010.        ],
       [3582.68736772, 2007.        ],
       [2231.8081416 , 2011.        ],
       [2659.43145076, 2016.        ]])

In [17]:
type(X)

numpy.ndarray

In [18]:
asia_df = df[df['origin'] == 'Asia']
df_sub = asia_df[['vehicle_weight', 'model_year']]
df7 = df_sub.head(7)
X = df7.to_numpy()

In [19]:
XTX = np.dot(X.T, X)

In [20]:
XTX_inv = np.linalg.inv(XTX)

In [21]:
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])

In [22]:
w = XTX_inv @ X.T @ y
w.sum()

np.float64(0.5187709081074016)